### PR TITLE
Avoid Notice: Trying to access array offset on value of type bool

### LINKED ISSUE #27774: Notice: Trying to access array offset on value of type bool
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

Any module or function that uses the native core PrestaShop function "isDiscounted" from the Product class will rise a Notice because an uncontrolled condition inside that function that rely that you have a cart already, but in many cases you will not.

image

So, the simple thing is to make the condition with if "$ids" is not false (that is the return of evaluating the context cart id_address into the Address::getCountryAndState without really a cart)
image

Of course the "problem" appear when debug mode is on.

Anyway, this produce a lot of lines into the error_log.

### Expected behavior

Not any warning.

### Steps to reproduce

1.- Install a module or create an script that uses the "Product::isDiscounted" function and hooked/executed on a display front hook
1a.- For example https://addons.prestashop.com/es/informes-estadisticas/18739-pixel-plus-eventos-api-de-conversiones-catalogo.html or 
[wim_test27774.zip](https://github.com/PrestaShop/PrestaShop/files/8141510/wim_test27774.zip)

2.- Enable debug mode or tail your error_log

3.- Visit any product of your store without a cart (new visitor).

4.- Execution stopped with debug mode on or a lot of lines into the error log.

### PrestaShop version(s) where the bug happened
1.7.8.0, 1.7.8.2, 1.7.8.3

### PHP version(s) where the bug happened

7.4

### If your bug is related to a module, specify its name and its version

facebookconversiontrackingplus v2.3.3 or any module or script using  Product::isDiscounted

### PR Request ###
https://github.com/PrestaShop/PrestaShop/pull/27775

### PR DESCRIPTION
**Pull request by danidomen**
When you don't have a cart already yet, any module or script that uses this function raises a Notice: Trying to access array offset on value of type bool

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?          | develop
| Description?      | When you don't have a cart already yet, any module or script that uses this function raises a Notice: Trying to access array offset on value of type bool, because Address::getCountryAndState could return an array with the id_country, or if the cart is false, the address also and the function return a false, that is not correct evaluated into the next line (only take account that is returning an array)
| Type?           | bug fix
| Category         | FO
| BC breaks        |  no
| Deprecations     |  no
| Fixed ticket?    | Fixes #27774.
| How to test      | On the issue #27774 is explained how to test it.
| Possible impacts | None, with that simple behaviour is only applying more correctly the condition if the above function returns false when a cart doesn't have a address initialized.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27775)

### CODE REVIEW

**Comment by danidomen** [comment]
@matthieu-rolland @eternoendless @matks (idk if you are the correct people sorry), the test Sanity 8.0 is failing with error on the test itself because it cannot install PrestaShop (timeout due cannot find where the button to install is):

> 1) Install Prestashop
       should agree to terms and conditions and go to step 'System compatibility':
     page.waitForSelector: Timeout 10000ms exceeded.
=========================== logs ===========================
waiting for selector "#sheet_system h2" to be visible
============================================================
      at Install.waitForSelector (pages/commonPage.js:43:16)
      at Install.waitForVisibleSelector (pages/commonPage.js:[54](https://github.com/PrestaShop/PrestaShop/runs/5291150828?check_suite_focus=true#step:17:54):16)
      at Install.getTextContent (pages/commonPage.js:99:18)
      at Install.getStepTitle (pages/install/index.js:117:17)
      at Context.<anonymous> (campaigns/sanity/01_installShop/01_installShop.js:68:43)

**Comment by matks** [comment]
> @matthieu-rolland @eternoendless @matks (idk if you are the correct people sorry), the test Sanity 8.0 is failing with error on the test itself because it cannot install PrestaShop (timeout due cannot find where the button to install is):
> 
> > 1. Install Prestashop
> >    should agree to terms and conditions and go to step 'System compatibility':
> >    page.waitForSelector: Timeout 10000ms exceeded.
> >    =========================== logs ===========================
> >    waiting for selector "#sheet_system h2" to be visible
> >    ============================================================
> >    at Install.waitForSelector (pages/commonPage.js:43:16)
> >    at Install.waitForVisibleSelector (pages/commonPage.js:[54](https://github.com/PrestaShop/PrestaShop/runs/5291150828?check_suite_focus=true#step:17:54):16)
> >    at Install.getTextContent (pages/commonPage.js:99:18)
> >    at Install.getStepTitle (pages/install/index.js:117:17)
> >    at Context. (campaigns/sanity/01_installShop/01_installShop.js:68:43)

Thanks for the info ! this happens sometimes randomly so I restart the CI

File: classes/Product.php [hunk 1]
```diff
@@ -4090,7 +4090,7 @@ public static function isDiscounted($id_product, $quantity = 1, Context $context
 
         $id_currency = (int) $context->currency->id;
         $ids = Address::getCountryAndState((int) $context->cart->{Configuration::get('PS_TAX_ADDRESS_TYPE')});
-        $id_country = $ids['id_country'] ? (int) $ids['id_country'] : (int) Configuration::get('PS_COUNTRY_DEFAULT');
+        $id_country = ($ids && $ids['id_country']) ? (int) $ids['id_country'] : (int) Configuration::get('PS_COUNTRY_DEFAULT');
```
Comment (atomiix) [inline classes/Product.php]:
```suggestion
        $id_country = (int) ($ids['id_country'] ?? Configuration::get('PS_COUNTRY_DEFAULT'));
```

Comment (danidomen) [reply classes/Product.php]: Applied with the null coalescing operator, lovely. Thanks :)

**Review by jolelievre — DISMISSED** [review]
Thanks @danidomen

**Review by matthieu-rolland — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Review by tegbessou — DISMISSED** [review]

**Review by marwachelly — CHANGES_REQUESTED** [review]
Hello,

The [issue](https://github.com/PrestaShop/PrestaShop/issues/27774) is a regression. I think it should be fixed on the branch 1.7.8.x istead of develop. 
Could you please check?

Thanks,

**Comment by danidomen** [comment]
> Hello,
> 
> The [issue](https://github.com/PrestaShop/PrestaShop/issues/27774) is a regression. I think it should be fixed on the branch 1.7.8.x istead of develop. Could you please check?
> 
> Thanks,

Sorry @marwachelly what do you want that I change? I mean, for the label "Waiting for author"...

**Comment by marwachelly** [comment]
> > Hello,
> > The [issue](https://github.com/PrestaShop/PrestaShop/issues/27774) is a regression. I think it should be fixed on the branch 1.7.8.x istead of develop. Could you please check?
> > Thanks,
> 
> Sorry @marwachelly what do you want that I change? I mean, for the label "Waiting for author"...

Could you please change the base to 1.7.8.x branch instead of develop?

**Comment by danidomen** [comment]
> > > Hello,
> > > The [issue](https://github.com/PrestaShop/PrestaShop/issues/27774) is a regression. I think it should be fixed on the branch 1.7.8.x istead of develop. Could you please check?
> > > Thanks,
> > 
> > 
> > Sorry @marwachelly what do you want that I change? I mean, for the label "Waiting for author"...
> 
> Could you please change the base to 1.7.8.x branch instead of develop?

@marwachelly Changed the base directly here on GitHub editing with the "Edit" button the PR, idk if this is the way because now a lot of files are on the PR that I didn't touch. If not, please do you that or explain how to do it.

image

**Comment by eternoendless** [comment]
@danidomen you need to rebase your changes on top of the `1.7.8.x` branch before you can change the base branch on GitHub. This can't be done through the GitHub interface, you have to do it locally. 

If you aren't familiar with git rebase, there are tons of tutorials online. But given the small number of modifications in this PR, it might be easier to just start over and create a new PR based on `1.7.8.x`.

**Comment by danidomen** [comment]
@eternoendless yes, a rebase for this simple change is more work than start directly fresh from the 1.7.8.x, but how to merge with this actual PR? I mean, as you say, for something so simple and basic we need to take sooooo long time to test, reproduce, approve, review, retest, reapprove, etc. that I don't want to lost at least what we gain here and start a fresh PR without "history" behind :D

I mean, not lose history for 21 days for a simple ?? addition to fix an uncontrolled array :P

**Comment by danidomen** [comment]
@eternoendless @marwachelly here you have it. Choose the one you want.
https://github.com/PrestaShop/PrestaShop/pull/27922

**Comment by eternoendless** [comment]
Thank you for the port @danidomen 

Nothing is lost, but since each PR is tested by QA, it's better just to test the one in `1.7.8.x`. 

The change will naturally be ported into the `develop` branch once #27922 has been merged, and `1.7.8.x` merged into `develop`. We can safely close this PR in favor of the other one.

**Pull request closed without merging**